doc/source/user/depending_on_numpy.rst

```diff
@@ -114,25 +114,18 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
 recommend all packages depending on NumPy to follow the recommendations in NEP
 29.
 
-For *run-time dependencies*, you specify the range of versions in
+For *run-time dependencies*, specify version bounds using
 ``install_requires`` in ``setup.py`` (assuming you use ``numpy.distutils`` or
-``setuptools`` to build). Getting the upper bound right for NumPy is slightly
-tricky. If we don't set any bound, a too-new version will be pulled in a few
-years down the line, and NumPy may have deprecated and removed some API that
-your package depended on by then. On the other hand if you set the upper bound
-to the newest already-released version, then as soon as a new NumPy version is
-released there will be no matching version of your package that works with it.
-
-What to do here depends on your release frequency. Given that NumPy releases
-come in a 6-monthly cadence and that features that get deprecated in NumPy
-should stay around for another two releases, a good upper bound is
-``<1.(xx+3).0`` - where ``xx`` is the minor version of the latest
-already-released NumPy. This is safe to do if you release at least once a year.
-If your own releases are much less frequent, you may set the upper bound a
-little further into the future - this is a trade-off between a future NumPy
-version _maybe_ removing something you rely on, and the upper bound being
-exceeded which _may_ lead to your package being hard to install in combination
-with other packages relying on the latest NumPy.
+``setuptools`` to build). The upper bound should be set to the next
+major release of NumPy, e.g. `numpy>=1.18,numpy<2`.
```

Comment:
I agree with this. It should be possible to separate the documenting of how versions are numbered with the recommending (if any) of how to pin upstream.

Reply:
NumPy 2.0 is the only release in which we are even contemplating making major breaking changes.  Otherwise, NumPy remains for the greater part backwards compatible, not counting ABI, bug-fixes, and small tweaks.  ABI breakage does not affect most Python packages.

That said, I don't consider this an important part of the PR, so I can remove that part, or adjust it if you have some guidance.

Reply:
> NumPy 2.0 is the only release in which we are even contemplating making major breaking changes

This is not actually true. Two examples that immediately come to mind which deprecated widely used functionality (which then breaks older releases of pretty much every major package, and not by accident):

- Indexing with floats errors: https://numpy.org/devdocs/release/1.12.0-notes.html#deprecationwarning-to-error
- (removal coming soon I think) deprecated `np.int`, `np.float` et al: https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated

I think there are more such large changes, and we're not always opposed to making them - it's a cost/benefit thing. Value-based casting is one we're looking at now for example.